doc/vmemcache.md

```diff
@@ -0,0 +1,187 @@
+
+*replacement_policy* may be:
+ * VMEMCACHE_REPLACEMENT_NONE: manual eviction only -- puts into a full
+   cache will fail
+ * VMEMCACHE_REPLACEMENT_LRU: least recently accessed entry will be evicted
+   to make space when needed
+
+
+void vmemcache_delete(VMEMcache *c);
+
+Frees any structures associated with the cache; there's no need to call
+this function on shutdown unless you intend to test your program against
+memory leaks.
+
+
+##### Use #####
+
+ssize_t vmemcache_get(VMEMcache *cache,
+	const void *key, size_t key_size,
+	void *vbuf, size_t vbufsize, size_t offset, size_t *vsize);
+
+Searches for an entry for the given *key*; it doesn't have to be
+zero-terminated or be text -- any sequence of bytes of length *key_size*
+is okay.  If found, the entry's value is copied to *vbuf* that has space
+for *vbufsize* bytes, optionally skipping *offset* bytes at the start.
+No matter if the copy was truncated or not, it's true size is stored into
+*vsize*; *vsize* remains unmodified if the key was not found.
+
+Return value is -1 on error, 0 if the key was not found, or number of bytes
+successfully copied otherwise.  Note that 0 may be legitimately returned
+even if the key was found (reading past end, zero-sized value, etc) --
+check *vsize* to tell that apart.
+
+
+int vmemcache_put(VMEMcache *cache,
+	const void *key, size_t key_size,
+	const void *value, size_t value_size);
+
+Inserts the given key:value pair into the cache.  Returns 0 on success,
+-1 on error.  Inserting a key that already exists will fail with EEXIST.
+
+
+int vmemcache_evict(VMEMcache *cache, const void *key, size_t ksize);
+
+Removes the given key from the cache.  If *key* is null and there is a
+replacement policy set, the oldest entry will be removed.  Returns 0 if
+an entry has been evicted, -1 otherwise.
+
+
+##### Callbacks #####
+
+You can register a hook to be called during eviction or after a cache miss,
+using ``vmemcache_callback_on_evict()`` or ``vmemcache_callback_on_miss()``,
+respectively.  The extra *arg* will be passed to your function.
+
+void vmemcache_on_evict(VMEMcache *cache,
+	const void *key, size_t key_size, void *arg);
+
+Called when an entry is being removed from the cache.  The eviction can't
+be prevented, but until the callback returns, the entry remains available
+for queries.  The thread that triggered the eviction is blocked in the
+meantime.
+
+
+int vmemcache_on_miss(VMEMcache *cache,
+	const void *key, size_t key_size, void *arg);
+
+Called when a *get* query fails, to provide an opportunity to insert the
+missing key.  If the callback returns zero, upon return the query will
+be retried (just once -- no call if it fails again).  Note that it's
+possible that the entry will be evicted between the insert and get.
+
+
+##### Misc #####
+
+int vmemcache_get_stat(VMEMcache *cache,
+	enum vmemcache_statistic stat,
+	void *value, size_t value_size);
+
+Obtains a piece of statistics about the cache.  The *stat* may be:
+ * VMEMCACHE_STAT_PUT
```

Comment:
It would be nice to bold those stats.
Use + list:
\+ \*\*VMEM_STAT_GET\*\* etc. (this and previous list)